llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp

```diff
@@ -3166,109 +3171,228 @@ bool SPIRVInstructionSelector::selectFirstBitHigh32(Register ResVReg,
       .constrainAllUses(TII, TRI, RBI);
 }
 
-bool SPIRVInstructionSelector::selectFirstBitHigh64(Register ResVReg,
-                                                    const SPIRVType *ResType,
-                                                    MachineInstr &I,
-                                                    bool IsSigned) const {
-  Register OpReg = I.getOperand(2).getReg();
-  // 1. split our int64 into 2 pieces using a bitcast
-  unsigned count = GR.getScalarOrVectorComponentCount(ResType);
-  SPIRVType *baseType = GR.retrieveScalarOrVectorIntType(ResType);
+bool SPIRVInstructionSelector::selectFirstBitSet64(
+    Register ResVReg, const SPIRVType *ResType, MachineInstr &I,
+    Register SrcReg, unsigned BitSetOpcode, bool SwapPrimarySide) const {
+  unsigned ComponentCount = GR.getScalarOrVectorComponentCount(ResType);
+  SPIRVType *BaseType = GR.retrieveScalarOrVectorIntType(ResType);
+  bool ZeroAsNull = STI.isOpenCLEnv();
+  Register ConstIntZero =
+      GR.getOrCreateConstInt(0, I, BaseType, TII, ZeroAsNull);
+  Register ConstIntOne =
+      GR.getOrCreateConstInt(1, I, BaseType, TII, ZeroAsNull);
+
+  // SPIRV doesn't support vectors with more than 4 components. Since the
+  // algoritm below converts i64 -> i32x2 and i64x4 -> i32x8 it can only
+  // operate on vectors with 2 or less components. When largers vectors are
+  // seen. Split them, recurse, then recombine them.
+  if (ComponentCount > 2) {
+    unsigned LeftComponentCount = ComponentCount / 2;
```

Comment:
I tried to write it to handle that case. It should just keep recursing and splitting the vectors in half until its under 2 components. 

I do think we are strictly limited by SPIRV here though. Say hlsl supported u64x8, we still have to accept the vec8 in as a parameter and return a vec8 out. Both of which would require invalid SPIRV

Reply:
Actually one thing I was considering was to just explicitly handle the vec3 and vec4 cases then assert for anything higher  but they are equally as messy while being strictly less general

Reply:
I think/wonder if its possible that a previous part of the code will force the vectors to be vec4 or smaller, but I'm unsure.

Reply:
> I tried to write it to handle that case. It should just keep recursing and splitting the vectors in half until its under 2 components.
> 
> I do think we are strictly limited by SPIRV here though. Say hlsl supported u64x8, we still have to accept the vec8 in as a parameter and return a vec8 out. Both of which would require invalid SPIRV

I don't think this code can validly handle vectors that are > vec4 because the splitting action will create vectors which are too large in some cases.

Reply:
Can you give an example? I'm pretty sure the splitting will never create a vector too large (but the merging back together certainly can)

Example:

Given `u64x12` the call stack becomes

```c++
selectFirstBitSet64(u64x12); // Top
selectFirstBitSet64Overflow(u64x12); // Top
  selectFirstBitSet64(u64x6); // Top.Left
  selectFirstBitSet64Overflow(u64x6);   // Top.Left
    selectFirstBitSet64(u64x3);  // Top.Left.Left
    selectFirstBitSet64Overflow(u64x3); // Top.Left.Left
      selectFirstBitSet64(u64);  // Top.Left.Left.Left
      selectFirstBitSet64(u64x2);  // Top.Left.Left.Right
    selectFirstBitSet64Overflow(u64x3); // Top.Left.Right
      selectFirstBitSet64(u64);  // Top.Left.Right.Left
      selectFirstBitSet64(u64x2);  // Top.Left.Right.Right
  selectFirstBitSet64Overflow(u64x6);   // Top.Right
    selectFirstBitSet64(u64x3);  // Top.Right.Left
    selectFirstBitSet64Overflow(u64x3); // Top.Right.Left
      selectFirstBitSet64(u64);  // Top.Right.Left.Left
      selectFirstBitSet64(u64x2);  // Top.Right.Left.Right
    selectFirstBitSet64Overflow(u64x3); // Top.Right.Right
      selectFirstBitSet64(u64);  // Top.Right.Right.Left
      selectFirstBitSet64(u64x2);  // Top.Right.Right.Right
```

Reply:
When you split the a size 12 vector for the potential recursive call, you create two intermediate registers which contain vectors of size 6. I think you can see this in your callstack actually.

Reply:
Ah yep. :/

I was hoping it was clean and only  `SrcReg` `ResReg`  were "bad" but nope.

Probably just go with the assert that its never larger than `u64x4` for now.